paddle/fluid/compiler/piano/note_builder.h

```diff
@@ -0,0 +1,124 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "glog/logging.h"
+#include "paddle/fluid/compiler/piano/note/opcode.h"
+#include "paddle/fluid/compiler/piano/shape.h"
+#include "paddle/fluid/platform/enforce.h"
+#include "paddle/fluid/platform/errors.h"
+
+namespace paddle {
+namespace piano {
+
+class NoteBuilder;
+
+// A Operand is generally constructed by a NoteBuilder, as the returned value of
+// an Instruction, and can be used to an operand for succeeding instructions
+class Operand {
+ public:
+  using ShapeClass = paddle::piano::Shape;
+
+  Operand() : instr_id_(-1), builder_(nullptr) {}
+
+  // whether this operand valid
+  bool Valid() const { return instr_id_ >= 0 && builder_ != nullptr; }
+
+  // the builder that contructs this operand
+  NoteBuilder* Builder() const {
+    PADDLE_ENFORCE_NE(builder_, nullptr,
+                      platform::errors::InvalidArgument("Builder is nullptr"));
+    return builder_;
+  }
+
+  // shape of this operand
+  const ShapeClass& Shape();
+
+ private:
+  // declare the folloing methods as private and only can be used from friend
+  // class, to prevent from illegal usage in complie-time
+  explicit Operand(NoteBuilder* builder) : instr_id_(-1), builder_(builder) {}
+  Operand(int64_t id, NoteBuilder* builder)
+      : instr_id_(id), builder_(builder) {}
+
+  int64_t Id() const { return instr_id_; }
+
+  friend class NoteBuilder;
+
+ private:
+  // the unique id that denotes which instruction generate this value
+  int64_t instr_id_;
+  // the builder that holds the instruction
+  NoteBuilder* builder_;
+};
+
+// A NoteBuilder keeps a list of instructions within the same Note function,
```

Comment:
`within the same Note function`? I think they are in the same Note module.

Reply:
Yes, same Module will be more precise